### PR TITLE
Allow to use @vue/component to set enable parsing vue objects

### LINKED ISSUE #109: How to make parser work on plain js files.
I looked into my projects and i found out one missing functionality here, sometimes we can create abstract components in js files with no template, with or without render function and in that case eslint is not going to work with this file because he don't know that this is a component.

**Solution:**
I propose to add a way to instruct parser to treat files with comment `/* @vue/component */` as vue components.

### PR DESCRIPTION
**Pull request by armano2**
fixes #109

### CODE REVIEW

File: tests/lib/rules/no-dupe-keys.js [hunk 1]
```diff
@@ -77,8 +77,8 @@ ruleTester.run('no-dupe-keys', rule, {
 
   invalid: [
     {
-      filename: 'test.vue',
-      code: `
+      // filename: 'test.vue',
```
Comment (michalsnik) [inline tests/lib/rules/no-dupe-keys.js]: Please create another tests for that, don't comment actual ones. Also try case with block comment too :)

Comment (armano2) [inline tests/lib/rules/no-dupe-keys.js]: ok

Comment (armano2) [inline tests/lib/rules/no-dupe-keys.js]: @michalsnik new tests added

**Comment by armano2** [comment]
@michalsnik new tests added

**Comment by armano2** [comment]
@michalsnik do you have any suggestions? I'm terrible at describing stuff xd

**Comment by michalsnik** [comment]
I was thinking about adding information in `Usage` section, that we currently detect components based on certain requirements (list them) and if someone wants custom `.js` files to be checked in terms of the available rules he/she needs to add special comment, so that we can assume it's also to be considered as a component.

Btw. Now I realised that there might be other cases, when you for example want to create few small components inside one `.js` files and compose them, you could get:
```
// utils/example.js
const ChildComponent = {
  template: `<div>{{msg}}</div>`,
  data () {
    return { msg: 'Hello child' }
  }
}

export default {
 template: `<child-component />`,
  components: {
    ChildComponent
  }
}
```

such a situation might also have place in `.vue` files. My proposition would be to detect if `@vue/component` is exactly one line above the `ObjectExpression` - if yes -> treat the `ObjectExpression` as a component. This should solve all cases and shouldn't be hard to implement :) What do you think @armano2 ?

The above file could look like this:
```
// @vue/component
const ChildComponent = {
  template: `<div>{{msg}}</div>`,
  data () {
    return { msg: 'Hello child' }
  }
}

export default {
 template: `<child-component />`,
  components: {
    ChildComponent
  }
}
```

**Comment by armano2** [comment]
hmm i like it, and i was thinking about it before.

component definition should work only if next block is
`export default {`,
`export const/var/let [name] = {`
`const/var/let [name] ={`

what do you think?

**Comment by armano2** [comment]
@michalsnik hmm i like it, and i was thinking about it before.

component definition should work only if next block is
`export default {`,
`export const/var/let [name] = {`
`const/var/let [name] = {`

what do you think?

**Comment by michalsnik** [comment]
There are also async components, I wonder if simple check for an object right below comment wouldn't be enough. Then it would be also possible to do:
```
Vue.component('async-example', function (resolve, reject) {
  setTimeout(function () {
    // @vue/component
    resolve({
      template: '<div>I am async!</div>'
    })
  }, 1000)
})
```

**Comment by armano2** [comment]
@michalsnik i'm going to create separate test file for all cases to have them all in one place and i'm going to include async component to

**Comment by armano2** [comment]
@michalsnik can you check if i didn't forgo about something?

https://github.com/armano2/eslint-plugin-vue/blob/patch-28-comment-vue-component/tests/lib/utils/vue-component.js

File: tests/lib/utils/vue-component.js [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+      parserOptions
+    },
+    {
+      filename: 'test.js',
+      code: `foo({ })`,
+      parserOptions
+    },
+    {
+      filename: 'test.js',
+      code: `Vue.component('async-example', function (resolve, reject) { })`,
```
Comment (michalsnik) [inline tests/lib/utils/vue-component.js]: `resolve({})` missing

File: tests/lib/utils/vue-component.js [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+      code: `export const foo = {}`,
+      parserOptions
+    },
+    {
+      filename: 'test.js',
+      code: `var foo = {}`,
+      parserOptions
+    },
+    {
+      filename: 'test.js',
```
Comment (michalsnik) [inline tests/lib/utils/vue-component.js]: Add some cases with `.jsx` and `.vue` files too that are not to be treated as components by default.

File: tests/lib/utils/vue-component.js [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+      errors: [makeError(2)]
+    },
+    {
+      filename: 'test.js',
+      code: `Vue.component({})`,
+      parserOptions,
+      errors: [makeError(1)]
+    },
+    {
+      filename: 'test.vue',
```
Comment (michalsnik) [inline tests/lib/utils/vue-component.js]: Add the same test for `test.jsx` file

File: tests/lib/utils/vue-component.js [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+      filename: 'test.js',
+      code: `Vue.component('async-example', function (resolve, reject) { })`,
+      parserOptions
+    }
+  ],
+  invalid: [
+    {
+      filename: 'test.js',
+      code: `Vue.component('async-example', function (resolve, reject) {
+        resolve({})
```
Comment (michalsnik) [inline tests/lib/utils/vue-component.js]: This should actually be not treated as component, we're not supporting this case by default, only with optional comment.

**Comment by armano2** [comment]
@michalsnik i applied all requested changes

File: tests/lib/utils/vue-component.js [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+const ruleTester = new RuleTester()
+ruleTester.run('vue-component', rule, {
+
+  valid: [
+    {
+      filename: 'test.js',
+      code: `export default { }`,
+      parserOptions
+    }
+  ].concat(notFoundTests('js')).concat(notFoundTests('jsx')).concat(notFoundTests('vue')),
```
Comment (michalsnik) [inline tests/lib/utils/vue-component.js]: I'd call it `validTests` and `invalidTests` for consistency.

File: tests/lib/utils/vue-component.js [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+      parserOptions
+    }
+  ]
+}
+
+function foundTests (ext) {
+  return [
+    {
+      filename: `test.${ext}`,
+      code: `// ${ext}
```
Comment (michalsnik) [inline tests/lib/utils/vue-component.js]: I guess you added those comments for easier debugging?
If yes please keep them in once place, as for now I can see them on top and bottom of the code example.

File: tests/lib/utils/vue-component.js [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+// ------------------------------------------------------------------------------
+// Requirements
+// ------------------------------------------------------------------------------
+
+const rule = {
+  create (context) {
+    return utils.executeOnVueComponent(context, obj => {
+      context.report({
+        node: obj,
+        message: 'Internal test: component.'
```
Comment (michalsnik) [inline tests/lib/utils/vue-component.js]: I think `Component detected` message would be a little bit more descriptive.

File: lib/utils/index.js [hunk 9]
```diff
@@ -432,22 +432,52 @@ module.exports = {
...
+      },
+      ObjectExpression (node) {
+        if (!componentComments.some(el => el.loc.end.line === node.loc.start.line - 1) || hasNode(node)) return
+        foundNodes.push(node)
+        cb(node)
+      },
       'ExportDefaultDeclaration:exit' (node) {
         // export default {} in .vue || .jsx
-        if (!_this.isVueComponentFile(node, filePath)) return
+        if (!_this.isVueComponentFile(node, filePath, context) || hasNode(node)) return
```
Comment (michalsnik) [inline lib/utils/index.js]: Why did you passed `context` to `isVueComponentFile`? And why did you added `hasNode` here? I thought current check is ok and we only needed to add comment detection before objects?

File: lib/utils/index.js [hunk 10]
```diff
@@ -432,22 +432,52 @@ module.exports = {
...
         cb(node.declaration)
       },
       'CallExpression:exit' (node) {
         // Vue.component('xxx', {}) || component('xxx', {})
-        if (!_this.isVueComponent(node)) return
+        if (!_this.isVueComponent(node) || foundNodes.some(el => el.loc.start.line === node.loc.start.line)) return
```
Comment (michalsnik) [inline lib/utils/index.js]: Why did you added here additional check too?

File: tests/lib/utils/vue-component.js [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+      filename: `test.${ext}`,
+      code: `// ${ext}
+      // @vue/component
+      export default { }`,
+      parserOptions,
+      errors: [makeError(3)]
+    },
+    {
+      filename: `test.${ext}`,
+      code: `// ${ext}
```
Comment (michalsnik) [inline tests/lib/utils/vue-component.js]: Please consider formatting those code example as follows:
```
code: `
  /* @vue/component */
  export default {}
`
```
I think they would be much easier to read :)

File: lib/utils/index.js (hunk 10 above)
Comment (armano2) [inline lib/utils/index.js]: to prevent duplicates when someone

File: lib/utils/index.js (hunk 9 above)
Comment (armano2) [inline lib/utils/index.js]: context removed and to prevent duplicates from callback

File: tests/lib/utils/vue-component.js (hunk 8 above)
Comment (armano2) [inline tests/lib/utils/vue-component.js]: @michalsnik changed

File: tests/lib/utils/vue-component.js (hunk 7 above)
Comment (armano2) [inline tests/lib/utils/vue-component.js]: Yes that was the point, and i moved them to down (i forgot to this will all of them)

**Comment by armano2** [comment]
@michalsnik thank you for suggestions, changes applied

File: lib/utils/index.js (hunk 10 above)
Comment (michalsnik) [inline lib/utils/index.js]: Ah I see, good thinking :)

**Comment by michalsnik** [comment]
Cool @armano2 Now please add info in readme and we're good to go :) Good work!

**Comment by armano2** [comment]
@michalsnik now the hardest part 👎

**Comment by armano2** [comment]
@michalsnik now the hardest part 👎

i wrote something, but i don't like it :(

**Comment by michalsnik** [comment]
@armano2 And what do you think about:

### Attention

All component-related rules are being applied to code that passes any of the following checks:
- `Vue.component()` expression
- `export default {}` in `.vue` or `.jsx` file

If you however want to take advantage of our rules in any of your custom objects that are Vue components, you might need to use special comment `// @vue/component` that marks object in the next line as a Vue component in any file, e.g.:

```
// @vue/component
const CustomComponent = {
  name: 'custom-component',
  template: '<div></div>'
}

Vue.component('AsyncComponent', (resolve, reject) => {
  setTimeout(() => {
    // @vue/component
    resolve({
      name: 'async-component',
      template: '<div></div>'
    })
  }, 500)
})
```

**Comment by armano2** [comment]
@michalsnik i like it way more

i'm just terrible at writing *documentations*

**Comment by armano2** [comment]
There is one issue with this solution, it will not work with eslint 4


https://github.com/eslint/eslint/blob/8796d55a75a2d54b384c2ce3249861d5fe127b12/docs/user-guide/migrating-to-4.0.0.md#-linecomment-and-blockcomment-events-will-no-longer-be-emitted-during-ast-traversal

**Comment by armano2** [comment]
@michalsnik  There is one issue with this solution, it will not work with eslint 4

https://github.com/eslint/eslint/blob/8796d55a75a2d54b384c2ce3249861d5fe127b12/docs/user-guide/migrating-to-4.0.0.md#-linecomment-and-blockcomment-events-will-no-longer-be-emitted-during-ast-traversal

**Pull request merged**